File: apps/server/src/schemas/profile-chat.ts
// Module: Defines TypeBox contracts for the profile collection chat endpoint.
import { t } from "elysia"
import type { Static } from "@sinclair/typebox"

export const REQUIRED_PROFILE_FIELDS = [
  "full_name",
  "location",
  "skills",
  "education",
] as const

export const ProfileDataSchema = t.Object({
  full_name: t.Nullable(t.String()),
  location: t.Nullable(t.String()),
  skills: t.Array(t.String()),
  projects: t.Array(t.String()),
  certifications: t.Array(t.String()),
  languages: t.Array(t.String()),
  work_experiences: t.Array(t.String()),
  education: t.Nullable(t.String()),
  github_url: t.Nullable(t.String()),
  additional_information: t.Nullable(t.String()),
})

export const ProfileChatRequestSchema = t.Object({
  message: t.String({ minLength: 1, maxLength: 4000 }),
  session_id: t.String({ minLength: 1, maxLength: 80, default: "default" }),
  source: t.Optional(
    t.Object({
      type: t.Literal("url"),
      value: t.String({ minLength: 1, maxLength: 2048 }),
    })
  ),
})

export const ProfileChatDocumentRequestSchema = t.Object({
  message: t.Optional(t.String({ maxLength: 4000 })),
  session_id: t.String({ minLength: 1, maxLength: 80, default: "default" }),
  document: t.File({
    maxSize: "5m",
  }),
})

export const ProfileChatResponseSchema = t.Object({
  reply: t.String(),
  session_id: t.String(),
  profile: ProfileDataSchema,
  missing_required_fields: t.Array(t.String()),
  is_profile_ready: t.Boolean(),
  redirect_to: t.Nullable(t.String()),
})

export type RequiredProfileField = (typeof REQUIRED_PROFILE_FIELDS)[number]
export type ProfileData = Static<typeof ProfileDataSchema>
export type ProfileChatRequest = Static<typeof ProfileChatRequestSchema>
export type ProfileChatDocumentRequest = Static<
  typeof ProfileChatDocumentRequestSchema
>
export type ProfileChatResponse = Static<typeof ProfileChatResponseSchema>
export type ProfileSourceType = "text" | "docx" | "url"
export type ProfileSourceContext = {
  type: ProfileSourceType
  label: string
  content: string
}
export type ProfilePatch = Partial<{
  full_name: string
  location: string
  skills: string[]
  projects: string[]
  certifications: string[]
  languages: string[]
  work_experiences: string[]
  education: string
  github_url: string
  additional_information: string
}>

File: apps/server/app/schemas/profile_chat.py
# Module: Defines Pydantic contracts for the profile collection chat agent.
from pydantic import BaseModel, Field, field_validator

REQUIRED_PROFILE_FIELDS = ["full_name", "location", "skills", "education"]


class ProfileData(BaseModel):
    full_name: str | None = None
    location: str | None = None
    skills: list[str] = Field(default_factory=list)
    projects: list[str] = Field(default_factory=list)
    certifications: list[str] = Field(default_factory=list)
    languages: list[str] = Field(default_factory=list)
    work_experiences: list[str] = Field(default_factory=list)
    education: str | None = None
    additional_information: str | None = None


class ProfileChatRequest(BaseModel):
    message: str = Field(min_length=1, max_length=4000)
    session_id: str = Field(default="default", min_length=1, max_length=80)

    @field_validator("message", "session_id")
    @classmethod
    def normalize_text(cls, value: str) -> str:
        return " ".join(value.strip().split())


class ProfileChatResponse(BaseModel):
    reply: str
    session_id: str
    profile: ProfileData
    missing_required_fields: list[str]
    is_profile_ready: bool
